pingcap/tidb-operator
# add local volume support for Backup/Restore using BR

### What problem does this PR solve? 
Add support for using local volumes (NFS or any fs that can be mounted on a local volume) as data backend when using BR to backup/restore data from TiDB clusters.

closes #3293

### What is changed and how does it work?
1. add `AdditionalVolumeMounts` in `ComponentSpec`, so that user can mount anything in `AdditionalVolumes` in the TiKV/PD/TiDB pods.
2. Backup/Restore reconciliation loop will check `Spec.Local`, and mount any volume specified there by user.
3. BR will start with `--storage local://` command line parameters to use the mounted volume, and backup to/restore from the volume.

### Check List 

Tests 

- Unit test
    - Manual test (add detailed scripts or steps below)
    - `Backup` CR
        - generate some data in the database (I used sysbench prepare subcommand)
        - create a `backup-nfs.yaml` file as follows, then use `kubectl apply` to apply the yaml
        - wait for the Job to complete, then the backup metadata and sst files will appear in the /nfs folder in the NFS server
        - run `kubectl delete -f backup-nfs.yaml` and notice that backup files are deleted because `spec.cleanPolicy` is set to `Delete`
        ```yaml
        # backup-nfs.yaml
        apiVersion: pingcap.com/v1alpha1
        kind: Backup
        metadata:
            name: basic-backup-nfs
            namespace: default
        spec:
            cleanPolicy: Delete
            br:
                cluster: basic
                clusterNamespace: default
            local:
                volume:
                name: vol1
                nfs:
                    server: 192.168.0.2
                    path: /nfs
                volumeMount:
                    name: vol1
                    mountPath: /vol1
        ```   
    - `Restore` CR
        - delete data from the database
        - create a `restore-nfs.yaml` file as follows, then use `kubectl apply` to apply the yaml. the location should point to the same nfs folder in the `Backup` CR
        - wait for the Job to complete, then the metadata and sst files will be restored into the database
        ```yaml
        # restore-nfs.yaml
        apiVersion: pingcap.com/v1alpha1
        kind: Restore
        metadata:
            name: basic-restore-nfs
            namespace: default
        spec:
            br:
                cluster: basic
                clusterNamespace: default
            local:
                volume:
                name: vol1
                nfs:
                    server: 192.168.0.2
                    path: /nfs
                volumeMount:
                    name: vol1
                    mountPath: /vol1
        ```
    - `BackupSchedule` CR
        - generate some data in the database (I used sysbench prepare subcommand)
        - create a `backup-schedule-nfs.yaml` file as follows, which schedules a backup job every minutes, then use `kubectl apply` to apply the yaml
        - wait for a few minutes and there should be multiple backups in `vol1`, which looks like `basic-pd.default-2379-2020-11-26t10-05-00`
        - run `kubectl delete -f backup-schedule-nfs.yaml` and notice that backup files are deleted because `spec.backupTemplate.cleanPolicy` is set to `Delete`
        ```yaml
        # backup-schedule-nfs.yaml
        apiVersion: pingcap.com/v1alpha1
        kind: BackupSchedule
        metadata:
            name: basic-backup-schedule-nfs
            namespace: default
        spec:
            maxReservedTime: "3h"
            schedule: "*/1 * * * *"
            backupTemplate:
                cleanPolicy: Delete
                br:
                    cluster: basic
                    clusterNamespace: default
                local:
                    volume:
                        name: vol1
                        nfs:
                            server: 192.168.0.2
                            path: /nfs
                    volumeMount:
                        name: vol1
                        mountPath: /vol1
        ```


Code changes

 - Has Go code change

Side effects

None

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support arbitrary volume and volumeMount for PD, TiDB, TiKV, TiFlash, Backup and Restore, which enables using NFS or any other kubernetes supported volume source for backup/restore workflow.
```
